### PR TITLE
Add start of state to CSRF cookie name

### PR DESCRIPTION
**Pull request by YoavOst**
## Description
This change adds a part of the state nonce to the CSRF name, making it possible to have parallel callback requests.

When having parallel ajax requests after the session cookie has expired multiple callbacks occur at the same time, creating multiple CSRF tokens with the name making it impossible for the proxy to authenticate.
It solves [issue 1451](https://github.com/oauth2-proxy/oauth2-proxy/issues/1451) and possibly [issue 817](https://github.com/oauth2-proxy/oauth2-proxy/issues/817)

## How Has This Been Tested?

Deployed using docker-compose with OIDC (keycloak) provider and cookie session store and solve the parallel AJAX requests issue.

### CODE REVIEW

File: pkg/cookies/csrf.go [hunk 1]
```diff
@@ -171,11 +177,11 @@ func decodeCSRFCookie(cookie *http.Cookie, opts *options.Cookie) (*csrf, error)
...
 // session cookie name
 func (c *csrf) cookieName() string {
-	return csrfCookieName(c.cookieOpts)
+	return csrfCookieName(c.cookieOpts, encryption.HashNonce(c.OAuthState)[0:csrfStateLength - 1])
 }
 
-func csrfCookieName(opts *options.Cookie) string {
-	return fmt.Sprintf("%v_csrf", opts.Name)
+func csrfCookieName(opts *options.Cookie, state string) string {
+	return fmt.Sprintf("%v_%v_csrf", opts.Name, state)
```
Comment (NickMeves) [inline pkg/cookies/csrf.go:184]: NIT -- can we potentially flip these so the dynamic part of the cookie name is the suffix? (`%v_csrf_%v`). And it might be worth renaming the `state` variable to signify this is a subset of the state.

Comment (NickMeves) [reply pkg/cookies/csrf.go:184]: In the CHANGELOG we will also need to signify this is potentially breaking for any users that had the cookieName option set close to the maximum (255 characters right?) -- Although I hope no one did that 🤣

File: pkg/cookies/csrf.go [hunk 2]
```diff
@@ -63,7 +66,10 @@ func NewCSRF(opts *options.Cookie) (CSRF, error) {
 
 // LoadCSRFCookie loads a CSRF object from a request's CSRF cookie
 func LoadCSRFCookie(req *http.Request, opts *options.Cookie) (CSRF, error) {
-	cookie, err := req.Cookie(csrfCookieName(opts))
+	
+	// csrfCookieName will include state to have multiple csrf in case of parallel requests
+	state := req.URL.Query()["state"][0][0:csrfStateLength - 1]
```
[the remark this answers is not in the document]
Comment (NickMeves) [reply pkg/cookies/csrf.go:71]: I'm also curious if we can potentially manage multiple CSRF state checks in an array structure serialized in 1 CSRF struct serialized into the cookie?

Things on my mind:

- The cookie manages the OIDC nonce across requests too. How will that handle multiple (seems fine).
- We encrypt the cookie payload (oauth state & nonce contents). Making the state part of the cookie name will slightly weaken that posture. But I bet it's fine, the encryption was likely overkill and those would only expose a portion of the OAuth state nonce.

[the remark this answers is not in the document]
Comment (NickMeves) [reply pkg/cookies/csrf.go:71]: I'm also not convinced the multiple tab issue was the sole source of these intermittent CSRF failures.

At my previous company where I ran this in production, whenever users reported this error (and when it happened to me) multiple tabs logging in a once was never a symptom.

[the remark this answers is not in the document]
Comment (NickMeves) [reply pkg/cookies/csrf.go:71]: This long running issue has more samples of tips people have (and theories on why this happens): https://github.com/oauth2-proxy/oauth2-proxy/issues/817

Do we think think this will fix some of the prefetch, favicon & JS issues people have uncovered as reasons?

**Comment by mazenaissa** [comment]
Hello @YoavOst ,
Could you please make the suggested changes? I am facing the 403 FORBIDDEN issue due to multiple oauth2 proxy callbacks invoked from multiple browser tabs.
Thanks,

File: pkg/cookies/csrf.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (YoavOst) [reply pkg/cookies/csrf.go]: Hi @JoelSpeed, Added some more testing for edge cases. 
@NickMeves I'm not sure about the multiple tabs, I have to say that from my experience, issues with multiple tabs can result from session management problems on the provider side.
I do suspect the parallel requests to favicon and prefetch are related to this.

File: pkg/cookies/csrf.go (hunk 1 above)
Comment (YoavOst) [reply pkg/cookies/csrf.go]: @NickMeves Sure. Fixed both

**Comment by BronzeDeer** [comment]
@NickMeves @YoavOst  Anyway other people can help get this reviewed and merged faster? At work we're also running into the parallel Ajax problem which is currently blocking

**Comment by mazenaissa** [comment]
> 

same here

**Comment by mazenaissa** [comment]
> @NickMeves @JoelSpeed @YoavOst Anyway other people can help get this reviewed and merged faster? At work we're also running into the parallel Ajax problem which is currently blocking

any updates guys?

**Review by JoelSpeed — COMMENTED** [review]
Has anyone tested this at all? Looking at the PR there are no test changes, can we make sure there are unit tests to ensure we don't regress on this functionality in the future?

**Comment by quantumdecoherence** [comment]
I've tested the PR, and can confirm that it solves the 403 FORBIDDEN problem. Thank you for that!

However, when the oauth2 session is expired, CSRF cookies are issued for every request made. The only CSRF cookie that's cleaned up is the one leading to a successful oauth2 login. This in turn leads to large headers containing all the uncleaned cookies being sent by the browser, and eventually to requests running into max header size.

**Comment by miguelborges99** [comment]
This fix tries to solve https://github.com/oauth2-proxy/oauth2-proxy/issues/1451 and possibly https://github.com/oauth2-proxy/oauth2-proxy/issues/817, which has more than 2 years. Can conflits be fixed?

**Review by JoelSpeed — APPROVED** [review]

**Comment by JoelSpeed** [comment]
Let's give this a go, hopefully this resolves the issues. It would be nice if someone could follow up with some unit tests for this

**Comment by JoelSpeed** [comment]
Ok this doesn't seem like it's working correctly as it's causing a panic within the tests, anyone got time to work out why and resolve that?

**Comment by mazenaissa** [comment]
Hello @JoelSpeed ,
Could you check this pull request https://github.com/oauth2-proxy/oauth2-proxy/pull/1708
One of the twos must be accepted to fix the CSRF issue
Thanks,

**Comment by JoelSpeed** [comment]
I'm going to close this in favour of #1708 as I think that's pretty much ready to go and seems to have had a bunch of testing from others
